### PR TITLE
Adding field for specifying possible public ip addresses of machine

### PR DESCRIPTION
**Pull request by preved911**
**What this PR does / why we need it**:
In some security cases we need to bind same ip addresses to new machines, when it was recreated. 

**Special notes for your reviewer**:
Added `publicIPAddresses` field into machineclass `providerSpec.networkInterfaceSpecs` items spec. Field accepts list of external ip addresses and change vm creation behaviour. VMs will be created only if one of the given external ip address exists in specified cloud provider account folder and it isn't in used state(address must be reserved previously).

**Release note**:
1. Added `publicIPAddresses` field into machineclass `providerSpec.networkInterfaceSpecs` items spec.

### CODE REVIEW

**Comment by ghost** [comment]
@preved911 Thank you for your contribution.

File: pkg/provider/core.go [hunk 1]
```diff
@@ -430,6 +441,28 @@ func FindInstanceByFolderAndName(ctx context.Context, sdk *ycsdk.SDK, folderID s
 	return result.Instances[0], nil
 }
 
+func FindUnusedAddressFromListByFolder(ctx context.Context, sdk *ycsdk.SDK, folderID string, addresses []string) (string, error) {
```
Comment (AxiomSamarth) [inline pkg/provider/core.go]: Minor nit. This function can be unexpected function.
```suggestion
func findUnusedAddressFromListByFolder(ctx context.Context, sdk *ycsdk.SDK, folderID string, addresses []string) (string, error) {
```

**Review by AxiomSamarth — CHANGES_REQUESTED** [review]

File: pkg/provider/core.go [hunk 2]
```diff
@@ -101,13 +103,22 @@ func (p *Provider) CreateMachine(ctx context.Context, req *driver.CreateMachineR
...
 	var networkInterfaceSpecs []*compute.NetworkInterfaceSpec
 	for _, netIf := range spec.NetworkInterfaceSpecs {
 		var natSpec *compute.OneToOneNatSpec
 		if netIf.AssignPublicIPAddress {
 			natSpec = &compute.OneToOneNatSpec{
 				IpVersion: compute.IpVersion_IPV4,
 			}
+
+			if len(netIf.PublicIPAddresses) > 0 {
+				natSpec.Address, err = FindUnusedAddressFromListByFolder(ctx, ySDK, folderId, netIf.PublicIPAddresses)
```
Comment (AxiomSamarth) [inline pkg/provider/core.go]: The function call should be updated now :P
```suggestion
				natSpec.Address, err = findUnusedAddressFromListByFolder(ctx, ySDK, folderId, netIf.PublicIPAddresses)
```

**Review by AxiomSamarth — CHANGES_REQUESTED** [review]

File: pkg/provider/core.go (hunk 2 above)
Comment (preved911) [reply pkg/provider/core.go]: I'm so sorry, my bug. Fixed now. Thanks.

**Review by AxiomSamarth — APPROVED** [review]
/needs-second-opinion

File: pkg/provider/core.go [hunk 3]
```diff
@@ -101,13 +103,22 @@ func (p *Provider) CreateMachine(ctx context.Context, req *driver.CreateMachineR
...
 		if netIf.AssignPublicIPAddress {
 			natSpec = &compute.OneToOneNatSpec{
 				IpVersion: compute.IpVersion_IPV4,
 			}
+
+			if len(netIf.PublicIPAddresses) > 0 {
+				natSpec.Address, err = findUnusedAddressFromListByFolder(ctx, ySDK, folderId, netIf.PublicIPAddresses)
+				if err != nil {
+					return nil, mcmstatus.Error(mcmcodes.InvalidArgument, err.Error())
+				}
```
Comment (himanshu-kun) [inline pkg/provider/core.go]: we can use [`NotFound`](https://github.com/gardener/machine-controller-manager/blob/78057547c9bb7208ca254943f0f263c16d661912/pkg/util/provider/machinecodes/codes/codes.go#L59) code , here as we can't consider the argument as invalid ( no spelling mistakes or wrong type)

File: pkg/provider/core.go [hunk 4]
```diff
@@ -430,6 +441,28 @@ func FindInstanceByFolderAndName(ctx context.Context, sdk *ycsdk.SDK, folderID s
...
+	result, err := sdk.VPC().Address().List(ctx, &vpc.ListAddressesRequest{
+		FolderId: folderID,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	for _, item := range result.Addresses {
+		if !item.GetUsed() {
+			external := item.GetExternalIpv4Address().GetAddress()
```
Comment (himanshu-kun) [inline pkg/provider/core.go:454]: we can add `if external==""` then continue , to avoid the extra looping

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
Thanks for the PR @preved911 . 
Just some code optimization requested , otherwise lgtm

File: pkg/provider/core.go (hunk 4 above)
Comment (preved911) [reply pkg/provider/core.go:454]: Please check again

**Review by himanshu-kun — APPROVED** [review]

File: pkg/provider/core.go [hunk 5]
```diff
@@ -430,6 +441,29 @@ func FindInstanceByFolderAndName(ctx context.Context, sdk *ycsdk.SDK, folderID s
...
+			if external := item.GetExternalIpv4Address().GetAddress(); external != "" {
+				for _, address := range addresses {
+					if external == address {
+						return external, nil
+					}
+				}
+			}
+		}
+	}
+
```
Comment (AxiomSamarth) [inline pkg/provider/core.go]: Though this addresses, previous comment from @himanshu-kun, I think this new implementation introduces higher depth nesting of `for` and `if` blocks which may not be a good thing from readability perspective. May be here's an alternative suggestion (please indent it accordingly). 

```suggestion
for _, item := range result.Addresses{
	if !item.GetUsed(){
		external := item.GetExternalIpv4Address().GetAddress()
		if external == "" {
			continue
		}
		for _, address := range addresses {
			if external == address {
				return external, nil
			}
		}
	}
}
```

**Review by AxiomSamarth — CHANGES_REQUESTED** [review]

File: pkg/provider/core.go (hunk 5 above)
Comment (preved911) [reply pkg/provider/core.go]: Thanks for explanation. Changed like this.

**Review by AxiomSamarth — APPROVED** [review]
/lgtm

**Comment by AxiomSamarth** [comment]
@preved911 Looks good to us. If your tests are succeeding with the latest changes then please feel free to merge the changes. (I believe you are the one of the maintainers of MCM-Provider-Yandex).

**Comment by preved911** [comment]
> @preved911 Looks good to us. If your tests are succeeding with the latest changes then please feel free to merge the changes. (I believe you are the one of the maintainers of MCM-Provider-Yandex).

I tested latest commit in my cases and it works well right now. But I don't have permission for merging and cannot do it self.

**Pull request merged**